### ISSUE TITLE
Explore French fine-tuning over English model

### ISSUE THREAD

**Issue by lissyx** [mozilla/DeepSpeech#1576]
Thanks for @nicolaspanel's work, there is a dataset of roughly 200 hours already available that we should be able to use: https://github.com/nicolaspanel/TrainingSpeech

It'd be great to:
 - write an importer for that, so one can use this as a datasource in the dataset mix
 - try and experiment with fine-tuning of the English model with those french data

**Comment by nicolaspanel** [comment]
thanks @lissyx 
To be precise, it is "only" 100h so far.
Still working on it to find/validate more sources and diversity.

**Comment by nicolaspanel** [comment]
@lissyx @kdavis-mozilla I can help for the importer.
Do you have a working exemple I can inspire from ?
Best regards

**Comment by lissyx** [comment]
Maybe `bin/import_cv.py` or others `import_*` scripts?

**Comment by lissyx** [comment]
@nicolaspanel I'm not sure 100%, but I ran into some errors that required me to set `ignore_longer_outputs_than_inputs=True` when calling `ctc_loss`. According do the documentation, it means we have longer sequence in output than in input. I'm trying fine-tuning over english model (so, changed the data to remove any accent and be able to use english alphabet file).

**Comment by lissyx** [comment]
First tentative, not so bad, a few hours of training on my GTX1080:
```
 python -u DeepSpeech.py --decoder_library_path /home/alexandre/tmp/deepspeech-0.2.1/libctc_decoder_with_kenlm.so --alphabet_config_path data/alphabet.txt  --train_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_train.csv   --dev_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_dev.csv   --test_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_test.csv   --train_batch_size 64   --dev_batch_size 64   --test_batch_size 64   --n_hidden 2048 --epoch 5 --learning_rate 0.0001 --display_step 0 --validation_step 1 --dropout_rate 0.2 --checkpoint_step 1 --checkpoint_dir ~/tmp/deepspeech-fr/checkpoints/ --export_dir ~/tmp/deepspeech-fr/model/ --initialize_from_frozen_model ~/tmp/deepspeech-0.2/models/output_graph.pb
```

And then a quick test:
```
$ time LD_LIBRARY_PATH=$(pwd)/../../tensorflow-lissyx/bazel-bin/native_client/ ./deepspeech --model ~/tmp/deepspeech-fr/model/output_graph.pb --alphabet ~/tmp/deepspeech-0.2/models/alphabet.txt --audio ~/tmp/deepspeech-fr/test-alex.fr.wav -t
TensorFlow: v1.11.0-5-g4cecc8c05b
DeepSpeech: v0.2.1-alpha.1-6-g72eaa2e
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2018-10-03 15:57:43.878247: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
bou je men sei moi
cpu_time_overall=8.43263
real    0m7,157s
user    0m4,838s
sys     0m3,918s
```

My audio is quite noisy, and I was saying "bonjour c'est moi"

**Comment by nicolaspanel** [comment]
> First tentative, not so bad, a few hours of training on my GTX1080:
> ```
>  python -u DeepSpeech.py --decoder_library_path /home/alexandre/tmp/deepspeech-0.2.1/libctc_decoder_with_kenlm.so --alphabet_config_path data/alphabet.txt  --train_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_train.csv   --dev_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_dev.csv   --test_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_test.csv   --train_batch_size 64   --dev_batch_size 64   --test_batch_size 64   --n_hidden 2048 --epoch 5 --learning_rate 0.0001 --display_step 0 --validation_step 1 --dropout_rate 0.2 --checkpoint_step 1 --checkpoint_dir ~/tmp/deepspeech-fr/checkpoints/ --export_dir ~/tmp/deepspeech-fr/model/ --initialize_from_frozen_model ~/tmp/deepspeech-0.2/models/output_graph.pb
> ```

Nice, thanks @lissyx  for sharing !

Which Language Model did you use ?

Do you have a first WER on the test set ?

NB: I noticed some errors on `2018-09-28_fr_FR`. You should experiment on `2018-10-03_fr_FR` which also contains more training data (see https://github.com/nicolaspanel/TrainingSpeech#last-releases--download)

**Comment by lissyx** [comment]
> > First tentative, not so bad, a few hours of training on my GTX1080:
> > ```
> >  python -u DeepSpeech.py --decoder_library_path /home/alexandre/tmp/deepspeech-0.2.1/libctc_decoder_with_kenlm.so --alphabet_config_path data/alphabet.txt  --train_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_train.csv   --dev_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_dev.csv   --test_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_test.csv   --train_batch_size 64   --dev_batch_size 64   --test_batch_size 64   --n_hidden 2048 --epoch 5 --learning_rate 0.0001 --display_step 0 --validation_step 1 --dropout_rate 0.2 --checkpoint_step 1 --checkpoint_dir ~/tmp/deepspeech-fr/checkpoints/ --export_dir ~/tmp/deepspeech-fr/model/ --initialize_from_frozen_model ~/tmp/deepspeech-0.2/models/output_graph.pb
> > ```
> 
> Nice, thanks @lissyx for sharing !
> 
> Which Language Model did you use ?

None so far, we'd need to think and build one for sure, but given that we need to change the hyperparameters of the current one I thought it'd be a good enough test :)

> 
> Do you have a first WER on the test set ?

```
I Test of Epoch 5 - WER: 0.403553, loss: 26.674508241089907, mean edit distance: 0.164999
I --------------------------------------------------------------------------------
I WER: 0.928571, loss: 0.000000, mean edit distance: 0.846154
I  - src: "allons dit le comte je vois que vous etes toujours le meme monsieur l'assassin"
I  - res: "monsieur l'abbe "
I --------------------------------------------------------------------------------
I WER: 0.928571, loss: 0.000000, mean edit distance: 0.846154
I  - src: "allons dit le comte je vois que vous etes toujours le meme monsieur l'assassin"
I  - res: "monsieur l'abbe "
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.000000, mean edit distance: 1.000000
I  - src: "et pardieu"
I  - res: ""
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.000000, mean edit distance: 1.000000
I  - src: "et pardieu"
I  - res: ""
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.048749, mean edit distance: 0.333333
I  - src: "non"
I  - res: "no "
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.048749, mean edit distance: 0.333333
I  - src: "non"
I  - res: "no "
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.085027, mean edit distance: 0.333333
I  - src: "non"
I  - res: "no "
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.085027, mean edit distance: 0.333333
I  - src: "non"
I  - res: "no "
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.242516, mean edit distance: 1.000000
I  - src: "oui"
I  - res: ""
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 0.242516, mean edit distance: 1.000000
I  - src: "oui"
I  - res: ""
I --------------------------------------------------------------------------------
```

> 
> NB: I noticed some errors on `2018-09-28_fr_FR`. You should experiment on `2018-10-03_fr_FR` which also contains more training data (see https://github.com/nicolaspanel/TrainingSpeech#last-releases--download)

Thanks

**Comment by lissyx** [comment]
> > None so far, we'd need to think and build one for sure, but given that we need to change the hyperparameters of the current one I thought it'd be a good enough test :)
> 
> +1
> 
> Have you heard of http://nlp.fast.ai/classification/2018/05/15/introducting-ulmfit.html ?

Interesting, but the language model itself is not a big big issue to deal with, I guess. Seems like we could learn a lot for other tasks of transfer learning, though. Might be useful to share that to people on Discourse that tries to do transfer :)

**Comment by lissyx** [comment]
> First tentative, not so bad, a few hours of training on my GTX1080:
> 
> ```
>  python -u DeepSpeech.py --decoder_library_path /home/alexandre/tmp/deepspeech-0.2.1/libctc_decoder_with_kenlm.so --alphabet_config_path data/alphabet.txt  --train_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_train.csv   --dev_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_dev.csv   --test_files /home/alexandre/tmp/deepspeech-fr/data/training_speech_2018-09-28_fr_FR_test.csv   --train_batch_size 64   --dev_batch_size 64   --test_batch_size 64   --n_hidden 2048 --epoch 5 --learning_rate 0.0001 --display_step 0 --validation_step 1 --dropout_rate 0.2 --checkpoint_step 1 --checkpoint_dir ~/tmp/deepspeech-fr/checkpoints/ --export_dir ~/tmp/deepspeech-fr/model/ --initialize_from_frozen_model ~/tmp/deepspeech-0.2/models/output_graph.pb
> ```
> 
> And then a quick test:
> 
> ```
> $ time LD_LIBRARY_PATH=$(pwd)/../../tensorflow-lissyx/bazel-bin/native_client/ ./deepspeech --model ~/tmp/deepspeech-fr/model/output_graph.pb --alphabet ~/tmp/deepspeech-0.2/models/alphabet.txt --audio ~/tmp/deepspeech-fr/test-alex.fr.wav -t
> TensorFlow: v1.11.0-5-g4cecc8c05b
> DeepSpeech: v0.2.1-alpha.1-6-g72eaa2e
> Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
> 2018-10-03 15:57:43.878247: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
> bou je men sei moi
> cpu_time_overall=8.43263
> real    0m7,157s
> user    0m4,838s
> sys     0m3,918s
> ```
> 
> My audio is quite noisy, and I was saying "bonjour c'est moi"

I've did some experiment, starting with english model 0.3.0, and @nicolaspanel https://gitlab.com/nicolaspanel/TrainingSpeech latest release (early october?) and relying on the newer `ctcdecode`, the same audio file gives "bonjour ses moi", without any language model, so that's quite accurate.

**Comment by lissyx** [comment]
```
$ LD_LIBRARY_PATH=$(pwd)/../../tensorflow-lissyx/bazel-bin/native_client/ ./deepspeech --model ~/tmp/deepspeech-fr/model/output_graph.pb --alphabet ~/tmp/deepspeech-0.3/models/alphabet.txt --audio ~/tmp/deepspeech-fr/test-alex.fr.wav -t
TensorFlow: v1.12.0-rc2-5-g1c93ca24c9
DeepSpeech: v0.4.0-alpha.0-0-g8b0abd5
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
bonjour ses moi
cpu_time_overall=8.58177
```

**Comment by elpimous** [comment]
top !!! very good !!

**Comment by lissyx** [comment]
After some experiments ran by @JRMeyer, turns out that for now, transfer learning, because of the amount of french data, is worse than training from scratch.

**Comment by nicolaspanel** [comment]
thanks @lissyx  for sharing
@JRMeyer  Did you use the same (ie english) alphabet for both experiments ?

**Comment by lissyx** [comment]
Work has resumed and is progressing on https://github.com/Common-Voice/commonvoice-fr/. Please check the Docker written for reproductibility https://github.com/Common-Voice/commonvoice-fr/blob/master/DeepSpeech/CONTRIBUTING.md as well as repo's releases.
